test/matrix.lean

```diff
@@ -0,0 +1,176 @@
+/-
+manually ported from
+https://github.com/leanprover-community/mathlib/blob/4f4a1c875d0baa92ab5d92f3fb1bb258ad9f3e5b/test/matrix.lean
+-/
+import Mathlib.Data.Matrix.Notation
+-- import linear_algebra.Matrix.determinant
+import Mathlib.GroupTheory.Perm.Fin
+-- import Mathlib.Tactic.NormSwap
+import Qq
+import Std.Tactic.GuardExpr
+
+open Qq
+
+-- todo: uncomment above imports when they are ported
+
+variables {α β : Type} [Semiring α] [Ring β]
+
+namespace Matrix
+
+open Matrix
+
+/-! Test that the dimensions are inferred correctly, even for empty matrices -/
+section dimensions
+
+-- set_option pp.universes true
+-- set_option pp.all true
+
+section elaborators
+open Lean Meta Elab Command
+
+/-- `dims% e` elaborates `e` as a Matrix and returns its dimensions as a `Nat × Nat`. -/
+elab "dims% " e:term : term => do
+  let elem_t ← mkFreshExprMVar (mkSort levelOne)
+  let m ← mkFreshExprMVar (mkConst ``Nat)
+  let n ← mkFreshExprMVar (mkConst ``Nat)
+  let matrix_t := mkAppN (mkConst ``Matrix [levelZero, levelZero, levelZero])
+                    #[mkApp (mkConst ``Fin) m, mkApp (mkConst ``Fin) n, elem_t]
+  let _ ← Term.elabTermEnsuringType e (some matrix_t)
+  let m ← instantiateMVars m
+  let n ← instantiateMVars n
+  mkAppM ``Prod.mk #[m, n]
+
+/-- Check equality of two expressions.
+* `#guard_expr e = e'` checks that `e` and `e'` are defeq at reducible transparency.
+* `#guard_expr e =~ e'` checks that `e` and `e'` are defeq at default transparency.
+* `#guard_expr e =ₛ e'` checks that `e` and `e'` are syntactically equal.
+* `#guard_expr e =ₐ e'` checks that `e` and `e'` are alpha-equivalent.
+
+This is similar to the `guard_expr` tactic but also uses default instances and instantiates
+metavariables. -/
```

Comment:
I don't like this slight difference in behavior. These little things are exactly why there are so many variations on guard_expr in the first place; there were 3 or 4 different functions, each one adding another delta on the last. If you need this behavior just PR it to std and make things consistent.

Reply:
I don't either, and truth be told that's why I left a note. I think the `guard_expr` tactic should also do metavariable instantiation and also make sure to use `synthesizeSyntheticMVarsUsingDefault`. (Perhaps `MatchKind.isEq` should be responsible for instantiating metavariables?)

This also isn't a simple matter of making a PR to Std to make things consistent: Std has a `guard_expr` tactic elaborator, but this is a `guard_expr#` command elaborator. I'd be happy if someone made the PR to include the new elaborator and make things consistent (and ideally even factoring out the body of both elaborators). Maybe I will do this eventually, but it's not high priority for me, and I'm happy enough with the partial progress toward improving `test/matrix.lean`.

Reply:
Could `#guard_expr` at least be split out as a separate PR? It's fantastic to have, but I think less confusing if it arrives in a separate PR/commit. I'll promise to merge quickly. :-)

Reply:
This PR is holding up progress on many fronts. Let's merge this and clean up `guard_expr` later.